cedar-policy-validator/src/schema.rs

```diff
@@ -475,48 +475,50 @@ impl ValidatorSchema {
     }
 
-    /// Get the validator entity equal to an EUID using the component for a head
-    /// var kind.
-    pub(crate) fn get_entity_eq<'a, H, K>(&self, var: H, euid: EntityUID) -> Option<K>
-    where
-        H: 'a + HeadVar<K>,
-        K: 'a,
-    {
-        var.get_euid_component(euid)
-    }
-
-    /// Get the validator entities that are in the descendants of an EUID using
-    /// the component for a head var kind.
-    pub(crate) fn get_entities_in<'a, H, K>(
+    /// Get all entity types in the schema which can be descendants of the given
+    /// entity type in the entity type graph.
+    pub(crate) fn get_entity_types_in<'a>(
         &'a self,
-        var: H,
-        euid: EntityUID,
-    ) -> impl Iterator<Item = K> + 'a
-    where
-        H: 'a + HeadVar<K>,
-        K: 'a + Clone,
-    {
-        var.get_descendants_if_present(self, euid.clone())
+        entity: &'a EntityUID,
+    ) -> impl Iterator<Item = &Name> + 'a {
+        let ety = match entity.entity_type() {
+            EntityType::Concrete(ety) => Some(ety),
+            EntityType::Unspecified => None,
+        };
+
+        ety.and_then(|ety| self.get_entity_type(ety))
             .into_iter()
-            .flatten()
-            .map(Clone::clone)
-            .chain(var.get_euid_component_if_present(self, euid))
+            .flat_map(|ety| ety.descendants.iter())
+            .chain(ety.filter(|ety| self.is_known_entity_type(ety)))
     }
 
-    /// Get the validator entities that are in the descendants of any of the
-    /// entities in a set of EUID using the component for a head var kind.
-    pub(crate) fn get_entities_in_set<'a, H, K>(
+    /// Get all entity types in the schema which can be a descendant of one of
+    /// the given entity types in the entity type graph.
+    pub(crate) fn get_entity_types_in_set<'a>(
         &'a self,
-        var: H,
-        euids: impl IntoIterator<Item = EntityUID> + 'a,
-    ) -> impl Iterator<Item = K> + 'a
-    where
-        H: 'a + HeadVar<K>,
-        K: 'a + Clone,
-    {
+        euids: impl IntoIterator<Item = &'a EntityUID> + 'a,
+    ) -> impl Iterator<Item = &Name> + 'a {
         euids
             .into_iter()
-            .flat_map(move |e| self.get_entities_in(var, e))
+            .flat_map(move |e| self.get_entity_types_in(e))
+    }
+
+    /// Get all action entities in the schema which are descendants of one of
+    /// the given entities in the action hierarchy.
+    pub(crate) fn get_actions_in_set<'a>(
+        &'a self,
+        euids: impl IntoIterator<Item = &'a EntityUID> + 'a,
+    ) -> impl Iterator<Item = &'a EntityUID> + 'a {
+        euids.into_iter().flat_map(|e| {
+            self.get_action_id(e)
+                .into_iter()
+                .flat_map(|action| action.descendants.iter())
+                .chain(if self.is_known_action_id(e) {
+                    Some(e)
+                } else {
+                    None
+                })
```

Comment:
does this imply that each action `e` is a descendant of itself?  (that kind of cycle isn't allowed by Core)